File: app/src/main/java/com/carwashpremiere/carwashpremieremobile/Act_DescriptionCars.java
package com.carwashpremiere.carwashpremieremobile;

import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.app.AppCompatDelegate;

import android.content.Context;
import android.content.Intent;
import android.content.res.Configuration;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.View;
import android.view.Window;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.AutoCompleteTextView;
import android.widget.Button;
import android.widget.Spinner;
import android.widget.TextView;
import android.widget.Toast;

import com.android.volley.Response;
import com.android.volley.VolleyError;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.List;

public class Act_DescriptionCars extends AppCompatActivity {
    Button btn_Next, btn_Back;
    TextView txt_ServiceDescription, txt_TitleCarServiceDetail;
    public static List<Model_CarTypes> mCarTypesList = new ArrayList<>();
    public static List<Model_ServicesCars> mServicesCarsList = new ArrayList<>();
    Function_NetworkRequests networkRequests;
    ArrayAdapter<Model_CarTypes> adapter;
    Model_CarTypes selectedAuto;
    String shortDescription, title, serviceName;
//    Spinner sp_CarType;

    AutoCompleteTextView autoCompleteTextView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        
        super.onCreate(savedInstanceState);
        requestWindowFeature(Window.FEATURE_NO_TITLE);
        getSupportActionBar().hide();

        setContentView(R.layout.activity_act_description_cars);

        txt_ServiceDescription = findViewById(R.id.txt_ServiceDescription);
//        sp_CarType = findViewById(R.id.sp_CarType);
        autoCompleteTextView = findViewById(R.id.autoComplete_CarType);
        btn_Next = findViewById(R.id.btn_NextObjectOrderDetail);
        btn_Back = findViewById(R.id.btn_Back);


        btn_Next.setVisibility(View.GONE);

        String[] types = new String[]{
                "toyota",
                "tsuru",
                "tesla"
        };


        Intent intent = getIntent();
        serviceName = intent.getStringExtra("serviceTitle");


        networkRequests = new Function_NetworkRequests(this);

        adapter = new ArrayAdapter<>(this, android.R.layout.simple_spinner_item, mCarTypesList);
        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);

//        sp_CarType.setAdapter(adapter);
        autoCompleteTextView.setAdapter(adapter);

        getCarTypesFromServer(networkRequests);
        getServicesCarsFromServer(networkRequests);

//        sp_CarType.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
//            @Override
//            public void onItemSelected(AdapterView<?> adapterView, View view, int position, long id) {
//
//                selectedAuto = (Model_CarTypes) adapterView.getItemAtPosition(position);
//
//
//                if (!selectedAuto.getTitle().equals("Seleccione una opción")) {
//
//                    btn_Next.setVisibility(View.VISIBLE);
//                } else {
//
//                    btn_Next.setVisibility(View.GONE);
//                }
//            }
//
//            @Override
//            public void onNothingSelected(AdapterView<?> adapterView) {
//                // No hacer nada si no se seleccionó ningún elemento
//            }
//        });

        // autoCompleteTextView.setAdapter(adapter);
        autoCompleteTextView.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {
                // No es necesario implementar esto, pero es requerido por el TextWatcher
            }

            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
                // Aquí obtienes el texto actualmente ingresado en el AutoCompleteTextView
                String inputText = charSequence.toString();

                // Buscas en la lista de tipos de automóviles si hay alguna coincidencia con el texto ingresado
                selectedAuto = null;
                for (Model_CarTypes carType : mCarTypesList) {
                    if (carType.getTitle().equals(inputText)) {
                        selectedAuto = carType;
                        break;
                    }
                }

                // Actualizas la visibilidad del botón según si se encontró una coincidencia válida
                if (selectedAuto != null && !selectedAuto.getTitle().equals("Seleccione una opción")) {
                    btn_Next.setVisibility(View.VISIBLE);
                } else {
                    btn_Next.setVisibility(View.GONE);
                }
            }

            @Override
            public void afterTextChanged(Editable editable) {
                // No es necesario implementar esto, pero es requerido por el TextWatcher
            }
        });


        btn_Back.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                // Terminar la actividad y regresar al menú principal

                Intent intent = new Intent(Act_DescriptionCars.this, Menu.class);
                intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);

                startActivity(intent);
                finish();
            }
        });



        btn_Next.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (selectedAuto != null) {
                    Intent intent = new Intent(Act_DescriptionCars.this, Act_CarParameters.class);
                    intent.putExtra("serviceTitle", serviceName);
                    intent.putExtra("carType", selectedAuto.getTitle());
                    startActivity(intent);
                } else {
                    Toast.makeText(Act_DescriptionCars.this, "Por favor, seleccione un tipo de automóvil.", Toast.LENGTH_SHORT).show();
                }
            }
        });




    }

    private void getCarTypesFromServer(Function_NetworkRequests networkRequests) {
        mCarTypesList.clear();
        mCarTypesList.add(new Model_CarTypes(1000, "Seleccione una opción"));
        networkRequests.getCarTypes(new Response.Listener<String>() {
            @Override
            public void onResponse(String response) {
                try {
                    JSONArray jsonArrayCategory = new JSONArray(response);

                    for (int i = 0; i < jsonArrayCategory.length(); i++) {
                        JSONObject jsonObject = jsonArrayCategory.getJSONObject(i);

                        int id = jsonObject.getInt("id");
                        String title = jsonObject.getString("title");


                        mCarTypesList.add(new Model_CarTypes(id, title));
                    }

                    adapter.notifyDataSetChanged();

                } catch (JSONException e) {
                    e.printStackTrace();
                }
            }
        }, new Response.ErrorListener() {

            @Override
            public void onErrorResponse(VolleyError error) {

                Toast.makeText(Act_DescriptionCars.this, "Error al recuperar información", Toast.LENGTH_LONG).show();
            }
        });
    }

    public void getServicesCarsFromServer(Function_NetworkRequests networkRequests) {
        mServicesCarsList.clear();
        networkRequests.getServicesCarDescription(new Response.Listener<String>() {
            @Override
            public void onResponse(String response) {
                try {
                    JSONObject jsonObject = new JSONObject(response);
                    shortDescription = jsonObject.getString("description");
                    title = jsonObject.getString("title");

                    txt_ServiceDescription.setText(shortDescription);
                } catch (JSONException e) {
                    e.printStackTrace();
                }
            }
        }, new Response.ErrorListener() {

            @Override
            public void onErrorResponse(VolleyError error) {
                Toast.makeText(Act_DescriptionCars.this, "Error al recuperar información: " + error.getMessage(), Toast.LENGTH_LONG).show();
            }
        }, serviceName);
    }



}

File: app/src/main/java/com/carwashpremiere/carwashpremieremobile/activities/Act_DescriptionCars.kt
package com.carwashpremiere.carwashpremieremobile.activities

import android.content.Intent
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.util.Log
import android.view.View
import android.view.Window
import android.widget.AdapterView
import android.widget.ArrayAdapter
import android.widget.AutoCompleteTextView
import android.widget.Button
import android.widget.Spinner
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.carwashpremiere.carwashpremieremobile.functions.Function_NetworkRequests
import com.carwashpremiere.carwashpremieremobile.R
import com.carwashpremiere.carwashpremieremobile.adapters.Adapter_CarTypeAdapter
import com.carwashpremiere.carwashpremieremobile.functions.Interface_RetrofitMethods
import com.carwashpremiere.carwashpremieremobile.functions.RetrofitClient
import com.carwashpremiere.carwashpremieremobile.model.Model_CarTypes
import com.carwashpremiere.carwashpremieremobile.model.Model_DetailsGeneral
import com.carwashpremiere.carwashpremieremobile.model.Model_ServicesCars
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class Act_DescriptionCars : AppCompatActivity() {
    var btn_Next: Button? = null
    var btn_Back: Button? = null
    var txt_ServiceDescription: TextView? = null
    var txt_title: TextView? = null
    var txt_TitleCarServiceDetail: TextView? = null
    var adapterCarType: Adapter_CarTypeAdapter? = null

    var adapter: ArrayAdapter<Model_CarTypes>? = null
    var selectedAuto: String? = null
    var sp_CarType: Spinner? = null
    var shortDescription: String? = null
    var title: String? = null
    var serviceName: String? = null

    //    Spinner sp_CarType;

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        requestWindowFeature(Window.FEATURE_NO_TITLE)
        supportActionBar!!.hide()

        setContentView(R.layout.activity_act_description_cars)

        InitUI()



        val intent = intent
        serviceName = intent.getStringExtra("serviceTitle")
        Log.e("Intent", serviceName.toString())
        getCarTypesSpinner()
        getService()



        btn_Back!!.setOnClickListener(View.OnClickListener { // Terminar la actividad y regresar al menú principal
            val intent = Intent(this@Act_DescriptionCars, Menu::class.java)
            intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP)
            startActivity(intent)
            finish()
        })

        btn_Next!!.setOnClickListener(View.OnClickListener {
            if (selectedAuto != null) {
                val intent = Intent(this@Act_DescriptionCars, Act_CarParameters::class.java)
                intent.putExtra("serviceTitle", serviceName)
                intent.putExtra("carType", selectedAuto!!)
                startActivity(intent)
            } else {
                Toast.makeText(
                    this@Act_DescriptionCars,
                    "Por favor, seleccione un tipo de automóvil.",
                    Toast.LENGTH_SHORT
                ).show()
            }
        })
    }

    fun InitUI(){
        txt_ServiceDescription = findViewById(R.id.txt_ServiceDescription)
        sp_CarType = findViewById(R.id.sp_CarType);
        txt_title = findViewById(R.id.txt_serviceCarTitle)

        btn_Next = findViewById(R.id.btn_NextObjectOrderDetail)
        btn_Back = findViewById(R.id.btn_Back)
        sp_CarType = findViewById(R.id.sp_CarType)
        btn_Next!!.setVisibility(View.GONE)

    }
    fun getService(){
        //TODO Assignate the specific api for service cars name
        val apiInterface = RetrofitClient.instance?.create(Interface_RetrofitMethods::class.java)
        val call = apiInterface?.getCarDescription_Name("11", serviceName.toString())

        call?.enqueue(object : Callback<Model_ServicesCars> {
            override fun onResponse(call: Call<Model_ServicesCars>, response: Response<Model_ServicesCars>){
                if (response.isSuccessful) {
                    val dataList = response.body()
                    if (dataList != null) {
                        txt_ServiceDescription?.text = dataList.description
                        txt_title?.text = dataList.title

                    }



                } else {
                    Toast.makeText(
                        applicationContext,
                        "Error al recuperar información: " + response.errorBody().toString(),
                        Toast.LENGTH_LONG
                    ).show()
                }
            }
            override fun onFailure(call: Call<Model_ServicesCars>, t: Throwable) {
                Toast.makeText(
                    applicationContext,
                    "Error al recuperar información: " + t.toString(),
                    Toast.LENGTH_LONG
                ).show()
            }
        })
    }
    fun getCarTypesSpinner(){
        val apiInterface = RetrofitClient.instance?.create(Interface_RetrofitMethods::class.java)
        val call = apiInterface?.getCarType()

        call?.enqueue(object : Callback<List<Model_CarTypes>> {
            override fun onResponse(call: Call<List<Model_CarTypes>>, response: Response<List<Model_CarTypes>>) {
                if (response.isSuccessful) {
                    val dataList = response.body()
                    if (dataList != null)
                    {
                        val carTypeTitles = dataList.map { it.title }
                        val adapter = ArrayAdapter(
                            this@Act_DescriptionCars,
                            android.R.layout.simple_spinner_item,
                            carTypeTitles
                        )
                        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
                        sp_CarType?.adapter = adapter

                        sp_CarType?.onItemSelectedListener =
                            object : AdapterView.OnItemSelectedListener {
                                override fun onItemSelected(
                                    adapterView: AdapterView<*>,
                                    view: View?,
                                    position: Int,
                                    id: Long
                                ) {
                                    selectedAuto =
                                        adapterView.getItemAtPosition(position) as String

                                    if (selectedAuto != "Seleccione una opción") {
                                        btn_Next?.visibility = View.VISIBLE
                                    } else {
                                        btn_Next?.visibility = View.GONE
                                    }
                                }

                                override fun onNothingSelected(adapterView: AdapterView<*>) {
                                    // No hacer nada si no se seleccionó ningún elemento
                                }
                            }
                    }



                } else {
                    // Manejar el error de la respuesta
                    Toast.makeText(
                        applicationContext,
                        "Error al recuperar información: " + response.errorBody().toString(),
                        Toast.LENGTH_LONG
                    ).show()
                }
            }

            override fun onFailure(call: Call<List<Model_CarTypes>>, t: Throwable) {
                Toast.makeText(
                    applicationContext,
                    "Error al recuperar información: " + t.message,
                    Toast.LENGTH_LONG
                ).show()
            }
        })
    }

}